### PR TITLE
Add configurable expiry time for local docs, defaults to 90 days

### LINKED ISSUE #1280: Auto-expire unused _local checkpoint documents
Sync Gateway doesn't currently set an expiry for _local documents.  This can result in an accumulation of obsolete replication checkpoint documents in the bucket.  

We can modify the _local document storage to:
(a) Set a relatively long expiry time on creation
(b) Update the expiry (GetAndTouch) when the checkpoint document is retrieved.

Clients that don't replicate within the expiry window will be forced to restart their replication from zero.  

The expiry value could default to something like 90 days, but probably makes sense as a config setting to let customers tune it to their use case.

### PR DESCRIPTION
**Pull request by ajres**
fixes #1280

### CODE REVIEW

File: rest/server_context.go [hunk 1]
```diff
@@ -284,6 +287,10 @@ func (sc *ServerContext) _getOrAddDatabaseFromConfig(config *DbConfig, useExisti
 		viewQueryTimeoutSecs = config.ViewQueryTimeoutSecs
 	}
 
+	if config.localDocExpirySecs != nil && *config.localDocExpirySecs > 0 {
```
Comment (adamcfraser) [inline rest/server_context.go]: I think explicitly setting config.localDocExpirySecs=0 should result in no expiry.  Here it seems to be reverting to the default value.

File: db/special_docs.go [hunk 2]
```diff
@@ -43,6 +54,10 @@ func (db *Database) putSpecial(doctype string, docid string, matchRev string, bo
 	if err != nil {
 		return "", base.HTTPErrorf(http.StatusBadRequest, "Invalid expiry: %v", err)
 	}
+
```
Comment (adamcfraser) [inline db/special_docs.go:57]: The current body.getExpiry() doesn't let us differentiate between _exp=0, and _exp not present.  It feels like we need that distinction in this case, to allow users to disable expiry per-doc.

File: db/special_docs.go [hunk 3]
```diff
@@ -24,10 +24,21 @@ func (db *Database) GetSpecial(doctype string, docid string) (Body, error) {
 	}
 
 	body := Body{}
-	_, err := db.Bucket.Get(key, &body)
-	if err != nil {
-		return nil, err
+
+	if doctype == "local" {
+		rawDocBytes, _, err := db.Bucket.GetAndTouchRaw(key, base.SecondsToCbsExpiry(int(db.DatabaseContext.Options.LocalDocExpirySecs)))
```
Comment (adamcfraser) [inline db/special_docs.go]: Can we do a regular Get if LocalDocExpirySecs is zero?

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: db/special_docs.go (hunk 3 above)
Comment (ajres) [reply db/special_docs.go]: Updated if condition to if doctype == "local" && db.DatabaseContext.Options.LocalDocExpirySecs > 0 {

File: rest/server_context.go (hunk 1 above)
Comment (ajres) [reply rest/server_context.go]: changed if condition to: 

 if config.localDocExpirySecs != nil && *config.localDocExpirySecs >= 0

File: db/special_docs.go (hunk 2 above)
Comment (ajres) [reply db/special_docs.go:57]: Added a bool return value 'present' to indicate that _exp was found in document

File: rest/config.go [hunk 4]
```diff
@@ -130,6 +130,7 @@ type DbConfig struct {
 	Unsupported          db.UnsupportedOptions          `json:"unsupported,omitempty"`                 // Config for unsupported features
 	OIDCConfig           *auth.OIDCOptions              `json:"oidc,omitempty"`                        // Config properties for OpenID Connect authentication
 	ViewQueryTimeoutSecs *uint32                        `json:"view_query_timeout_secs,omitempty"`     // The view query timeout in seconds
+	localDocExpirySecs   *uint32                        `json:"local_doc_expiry_secs,omitempty"`       //The _local doc expiry time in seconds
```
Comment (adamcfraser) [inline rest/config.go]: This should be exported (LocalDocExpirySecs), so that it's included in things like /db/_config

**Review by adamcfraser — CHANGES_REQUESTED** [review]
Caught one minor issue that I missed last time around - otherwise everything looks good.

File: rest/config.go (hunk 4 above)
Comment (ajres) [reply rest/config.go]: LocalDocExpirySecs has been exported

File: rest/server_context.go [hunk 5]
```diff
@@ -505,6 +512,7 @@ func (sc *ServerContext) _getOrAddDatabaseFromConfig(config *DbConfig, useExisti
 		IndexOptions:          channelIndexOptions,
 		SequenceHashOptions:   sequenceHashOptions,
 		RevisionCacheCapacity: revCacheSize,
+		LocalDocExpirySecs:	   localDocExpirySecs,
```
Comment (adamcfraser) [inline rest/server_context.go]: go fmt needed?

File: rest/server_context.go [hunk 6]
```diff
@@ -266,6 +266,9 @@ func (sc *ServerContext) _getOrAddDatabaseFromConfig(config *DbConfig, useExisti
 	pool := "default"
 	bucketName := config.Name
 
+	// Default value of _local document expiry
+	localDocExpirySecs := uint32(60 * 60 * 24 * 90) //90 days in seconds
```
Comment (adamcfraser) [inline rest/server_context.go]: Can you move this default value to base/constants.go to make it easier to find/maintain?

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: rest/server_context.go (hunk 6 above)
Comment (ajres) [reply rest/server_context.go]: Moved to base/constants.go and renamed to DefaultLocalDocExpirySecs

File: rest/server_context.go (hunk 5 above)
Comment (ajres) [reply rest/server_context.go]: reformatted

**Pull request merged**